rust/src/smb/smb.rs

```diff
@@ -0,0 +1,1862 @@
+            self.vercmd.set_status_dos_error(status);
+        } else {
+            self.vercmd.set_ntstatus(status);
+        }
+    }
+
+    pub fn free(&mut self) {
+        if self.events != std::ptr::null_mut() {
+            sc_app_layer_decoder_events_free_events(&mut self.events);
+        }
+        match self.de_state {
+            Some(state) => {
+                sc_detect_engine_state_free(state);
+            }
+            _ => {}
+        }
+    }
+}
+
+impl Drop for SMBTransaction {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+
+#[derive(Hash, Eq, PartialEq, Debug, Clone)]
+pub struct SMBFileGUIDOffset {
+    pub guid: Vec<u8>,
+    pub offset: u64,
+}
+
+impl SMBFileGUIDOffset {
+    pub fn new(guid: Vec<u8>, offset: u64) -> SMBFileGUIDOffset {
+        SMBFileGUIDOffset {
+            guid:guid,
+            offset:offset,
+        }
+    }
+}
+
+/// type values to make sure we're not mixing things
+/// up in hashmap lookups
+pub const SMBHDR_TYPE_GUID:        u32 = 1;
+pub const SMBHDR_TYPE_SHARE:       u32 = 2;
+pub const SMBHDR_TYPE_FILENAME:    u32 = 3;
+pub const SMBHDR_TYPE_OFFSET:      u32 = 4;
+pub const SMBHDR_TYPE_GENERICTX:   u32 = 5;
+pub const SMBHDR_TYPE_HEADER:      u32 = 6;
+pub const SMBHDR_TYPE_MAX_SIZE:    u32 = 7; // max resp size for SMB1_COMMAND_TRANS
+pub const SMBHDR_TYPE_TXNAME:      u32 = 8; // SMB1_COMMAND_TRANS tx_name
+pub const SMBHDR_TYPE_TRANS_FRAG:  u32 = 9;
+pub const SMBHDR_TYPE_TREE:        u32 = 10;
+pub const SMBHDR_TYPE_DCERPCTX:    u32 = 11;
+
+#[derive(Hash, Eq, PartialEq, Debug)]
+pub struct SMBCommonHdr {
+    pub ssn_id: u64,
+    pub tree_id: u32,
+    pub rec_type: u32,
+    pub msg_id: u64,
+}
+
+impl SMBCommonHdr {
+    pub fn init() -> SMBCommonHdr {
+        SMBCommonHdr {
+            rec_type : 0,
+            ssn_id : 0,
+            tree_id : 0,
+            msg_id : 0,
+        }
+    }
+    pub fn new(rec_type: u32, ssn_id: u64, tree_id: u32, msg_id: u64) -> SMBCommonHdr {
+        SMBCommonHdr {
+            rec_type : rec_type,
+            ssn_id : ssn_id,
+            tree_id : tree_id,
+            msg_id : msg_id,
+        }
+    }
+    pub fn from2(r: &Smb2Record, rec_type: u32) -> SMBCommonHdr {
+        let tree_id = match rec_type {
+            SMBHDR_TYPE_TREE => { 0 },
+            _ => r.tree_id,
+        };
+        let msg_id = match rec_type {
+            SMBHDR_TYPE_TRANS_FRAG => { 0 },
+            SMBHDR_TYPE_SHARE => { 0 },
+            _ => { r.message_id as u64 },
+        };
+
+        SMBCommonHdr {
+            rec_type : rec_type,
+            ssn_id : r.session_id,
+            tree_id : tree_id,
+            msg_id : msg_id,
+        }
+
+    }
+    pub fn from1(r: &SmbRecord, rec_type: u32) -> SMBCommonHdr {
+        let tree_id = match rec_type {
+            SMBHDR_TYPE_TREE => { 0 },
+            _ => r.tree_id as u32,
+        };
+        let msg_id = match rec_type {
+            SMBHDR_TYPE_TRANS_FRAG => { 0 },
```

Comment:
You can combine cases here:
`SMBHDR_TYPE_TRANS_FRAG |
SMBHDR_TYPE_SHARE => 0,`